### DISCUSSION TITLE
how to write tests?

### DISCUSSION THREAD

**Discussion by glehmann** [ebkalderon/tower-lsp#418] [Q&A]
Hi!

I’d like to write tests for my language server, in order to make sure to not break anything when adding new features or refactoring — just the usual good software development practices :)

It’s easy enough to do with unit tests for simple cases, but unfortunately I can’t find any example of a more complex example that requires several input files, require a full initialized backend or check that modified files or actually written in the expected place. The kind of integration test I expect to find in the `tests` directory.

Could someone point me to an example or a real project with this kind of test?

Gaëtan

**Comment by glehmann** [comment]
https://github.com/ast-grep/ast-grep/blob/main/crates/lsp/tests/basic.rs might be a good start, even if a bit rougher than I hoped.

It seems not many of the language server implemented in rust have this kind of test!

**Reply by estin** [reply]
similar  "rough" approach https://github.com/estin/simple-completion-language-server/blob/main/tests/basic.rs

**Reply by glehmann** [reply]
much less rough, with a test context in place of the input/output streams and a `recv` method to deserialize the message. I like it.

Thanks!

**Reply by glehmann** [reply]
here is my take on running integration tests in my tower-lsp based project

There is a module that is used in every test file: https://github.com/glehmann/earthlyls/blob/main/tests/common/mod.rs
A `TestContext` is created in the test that:
* creates a temp dir to be used as workspace
* copy some test files in the workspace
* instantiate a backend
The test can then send requests to the backend using lsp types and get the result back as lsp type

For example:

```rust
#[tokio::test]
async fn should_provide_references_with_wildcard() {
    let mut ctx = TestContext::new();
    ctx.initialize().await;
    let res = ctx
        .request::<request::References>(ReferenceParams {
            context: ReferenceContext { include_declaration: false },
            partial_result_params: PartialResultParams { partial_result_token: None },
            text_document_position: TextDocumentPositionParams {
                position: Position { line: 2, character: 3 },
                text_document: TextDocumentIdentifier { uri: ctx.doc_uri("bar/Earthfile") },
            },
            work_done_progress_params: WorkDoneProgressParams { work_done_token: None },
        })
        .await
        .unwrap();
    assert_eq!(res.len(), 1);
    let r = &res[0];
    assert_eq!(r.uri, ctx.doc_uri("Earthfile"));
    assert_eq!(r.range.start.line, 10);
    assert_eq!(r.range.start.character, 8);
    assert_eq!(r.range.end.line, 10);
    assert_eq!(r.range.end.character, 19);
}
```
The creation of the request param is a bit too verbose for me — I would have preferred to be able to use `Default::default()` to initialize the non required values. Not a big problem, I’m quite happy with the result :-)
